internal/pubsub/pubsub.go

```diff
@@ -197,12 +200,21 @@ func (s *Subscription) getNextBatch(ctx context.Context) error {
 }
 
 // Close flushes pending ack sends and disconnects the Subscription.
-func (s *Subscription) Close() error {
+func (s *Subscription) Shutdown(ctx context.Context) error {
+	c := make(chan struct{})
```

Comment:
I think this is all correct, but I find it hard to think about because the order in which things happen is complicated (first the function returns, then the channel gets closed, then the goroutine exits). I think the following is equivalent:
```
s.mu.Lock()
s.err = ...
s.mu.Unlock()
c := make(chan struct{})
go func() { defer close(c); s.ackBatcher.Shutdown() }
select {
case <-ctx.Done();
    s.cancel()
    return ctx.Err()
case <-c:
   return nil
}
```
except that it's easier to think about: "shut down the batcher, then wait for either that, or ctx.Done."

Reply:
One more thing, you're always supposed to call `cancel`, so I think the second select case should also say `s.cancel`. Or maybe:
```
...
select {
case <-ctx.Done():
case <-c:
}
s.cancel()
return ctx.Err()
```

Reply:
Done.